crates/bevy_ecs/src/storage/table.rs

```diff
@@ -262,31 +262,68 @@ impl Column {
     }
 }
 
-pub struct Table {
+/// A builder type for constructing [`Table`]s.
+///
+///  - Use [`with_capacity`] to initialize the builder.
+///  - Repeatedly call [`add_column`] to add columns for components.
+///  - Finalize with [`build`] to get the constructed [`Table`].
+///
+/// [`with_capacity`]: Self::with_capacity
+/// [`add_column`]: Self::add_column
+/// [`build`]: Self::build
+pub(crate) struct TableBuilder {
     columns: SparseSet<ComponentId, Column>,
-    entities: Vec<Entity>,
+    capacity: usize,
 }
 
-impl Table {
-    pub(crate) fn with_capacity(capacity: usize, column_capacity: usize) -> Table {
+impl TableBuilder {
+    /// Creates a blank [`Table`], allocating space for `column_capacity` columns
+    /// with the capacity to hold `capacity` entities worth of components each.
+    pub fn with_capacity(capacity: usize, column_capacity: usize) -> Self {
```

Comment:
Is there a chance that we end up realloc-ing our tables on construction in some / all cases, due to the fact that with_capacity isn't guaranteed to allocate exactly `capacity`, but `shrink_to_fit` will shrink to capacity?

Reply:
There is, but unless we're constantly making new `Tables` and `Archetypes`, the cost IMO is negligible long term.

Reply:
The frequency might go up drastically if we start doing "fragmenting relations". In that context the bytes saved might not be worth it. Hard to say what to value more.

I guess if/when we implement fragmenting relations we can let benchmarks decide. But this level of nuance is easy to forget about.